server/src/main/java/com/google/coffeehouse/common/Book.java

```diff
@@ -0,0 +1,159 @@
+
+package com.google.coffeehouse.common;
+
+import com.google.coffeehouse.util.IdentifierGenerator;
+import com.google.coffeehouse.util.UuidWrapper;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Book encapsulates the information associated with each book being used by a Club.
+ * 
+ * <p>Implementes the {@link Saveable} interface because it will be able to save itself
+ * in the database.
+ */
+public class Book implements Saveable {
+  /** The names of each possible field in the map in fromMap */
+  public static final String AUTHOR_FIELD_NAME = "author";
+  public static final String ISBN_FIELD_NAME = "isbn";
+  public static final String TITLE_FIELD_NAME = "title";
+
+  /** The message on the IllegalArgumentException when there is no "title" key */
+  public static final String NO_VALID_TITLE_FROM_MAP = 
+      "No valid \"" + TITLE_FIELD_NAME + "\" key defined.";
+  private String title;
+  private String author;
+  private String isbn;
+  private String bookId;
+
+  private Book(Builder builder) {
+    this.title = builder.title;
+    this.author = builder.author;
+    this.isbn = builder.isbn;
+    this.bookId = builder.idGenerator.generateId();
+  }
+
+  /** Overloaded static factory, calls other fromMap with null IdentifierGenerator. */
+  public static Book fromMap(Map bookInfo) {
+    return Book.fromMap(bookInfo, null);
+  }
+
+  /** 
+   * Creates a {@link Book} object from a Map with the relevant parameters as keys.
+   * @param bookInfo the Map that contains the information used to construct the Book. At a
+   *     minimum this includes a {@code "title"} key that is mapped to a String that describes
+   *     the title of the Book to be created. Optional keys include a {@code "author"} key 
+   *     that is mapped to a String describing the author of the book, as well as an 
+   *     {@code "isbn"} key that is mapped to a String that is the isbn number of the book
+   * @param idGen the {@link IdentifierGenerator} used when constructing the Book. If
+   *     null, the default generator will be used
+   * @return the created Book
+   * @throws IllegalArgumentException if no valid {@code "title"} key is defined
+   */
+  public static Book fromMap(Map bookInfo, IdentifierGenerator idGen) {
+    String title = (String) bookInfo.getOrDefault(TITLE_FIELD_NAME, null);
+    if (title == null) {
+      throw new IllegalArgumentException(NO_VALID_TITLE_FROM_MAP);
+    }
+
+    Book.Builder bookBuilder = Book.newBuilder(title);
+    if (bookInfo.containsKey(AUTHOR_FIELD_NAME)) {
+      bookBuilder.setAuthor((String) bookInfo.get(AUTHOR_FIELD_NAME));
+    }
+    if (bookInfo.containsKey(ISBN_FIELD_NAME)) {
+      bookBuilder.setIsbn((String) bookInfo.get(ISBN_FIELD_NAME));
+    }
+    if (idGen != null) {
+      bookBuilder.setIdGenerator(idGen);
+    }
+    return bookBuilder.build();
+  }
+
+  public String getTitle() {
+    return title;
+  }
+
+  /** Returns an optional because the author may, or may not, exist. */
```

Comment:
In the DB, I have author as not nullable. I'm assuming this comment means that I should change that, correct?

Reply:
Currently the author is an optional parameters in Book.java. If it is too hard to change the nullability of author in the database we can discuss solutions.

Reply:
I can just change it in the DB.